### DISCUSSION TITLE
Error Trapping

### DISCUSSION THREAD

**Discussion by Kr00l** [twinbasic/twinbasic#733] [Ideas]
This topic is intended to discuss general Error Trapping topics and how to include them in tB.
As you know VBx is terrible in this regard as it is defined by an IDE option setting ONLY and can not be controlled by code or certain project settings.

Thus, I guess tB want's to be better.

I think the Error Trapping needs to be similar in hierarchy like the Compiler Warnings. But with one extra layer.
**1.** Attributes which specify Error Trapping mode by code (module/class wide or certain methods only)
**2.** Project setting (default is "inherit from IDE setting")
**3.** IDE setting

At the end this will give most flexibility and can make everyone happy. Also those who want to continue working like in VBx.

To give a starter on the discussion I have following discrepancy discovered:

Having below sample code, tB breaks currently on MsgBox.
```
Class Class1
    
    Sub Test()
        MsgBox (1 / 0)
    End Sub
End Class

Module MainModule

    Public Sub Main()
        With New Class1
        .Test
        End With
    End Sub
	
End Module
```
This is the same as in VBx for option set "Break in Class Module"

When VBx is set to "Break on Unhandled Errors" the debugger breaks at the `.Test` line in Sub Main.

So the discrepancy is between VBx and tB if we set an OERN in Sub Main.
```
Public Sub Main()
    With New Class1
    On Error Resume Next
    .Test
    On Error GoTo 0
    End With
End Sub
```
tB currently runs without error message. However, in VBx when set to "Break in Class Module" it will bomb out at MsgBox in Class1.

So, that's little bit confusing to me.

**Comment by WaynePhillipsEA** [comment]
At the moment, the tB debugger is in a sort of _mixed_ mode.... "Break On Unhandled Errors AND in Class Modules".  This is what I consider to be a more natural way of debugging... since it more accurately reflects what happens when you compile to an EXE/DLL.   For example, if we compile your OERN example in VB6 to an EXE, the error does propagate back to Main() and the code does continue beyond the .Test call as expected without interruption... so the current tB debugging behaviour matches that.

That said, I do agree that we should still offer `Break on Unhandled Errors` and `Break in Class Modules` modes of course.

**Comment by WaynePhillipsEA** [comment]
My gut feeling is that we don't need the `IDE setting`.   In my opinion, the setting should be allowed via attributes in methods and components, and then at the project scope only.  Provided we can modify the project setting easily (e.g. via the twinBASIC panel), then I don't think we need the extra option at the IDE level.

I think the available options should be the usual `Break On All Errors`, `Break In Class Modules`, `Break On Unhandled Errors` plus a new `Break On Unhandled Errors (including Class Modules)` for the new tB offering.

**Reply by bclothier** [reply]
I'm inclined to agree that we don't need/want a IDE setting. I found that to be a weird thing.

That said, all those describe the debugger behavior, not runtime behavior. If I saw an attribute `[BreakOnAllErrors]`, it may read as modifying how it behaves at runtime, which would be wrong. I think that a lot of VBx users don't normally fiddle with the error settings so may not be aware of those behavior.

I'm thinking it will be clearer to decorate with something like:
```
[DebuggerBehavior(BreakOnAllErrors)]
[DebuggerBehavior(BreakInClassModules)]
[DebuggerBehavior(BreakOnUnhandledError)]
[DebuggerBehavior(MatchRuntimeBehavior)]
```

I think it also should be added that this should have the project scope. For example, if I import a twinpack with the `Break On All Errors`, it should not apply when I'm running my project which may have a different setting.

**Reply by Kr00l** [reply]
for me it's a little bit of learning at this topic as I didn't bother before.

So in fact the IDE setting in VBx doesn't influence the error behavior in the compiled runtime?
If yes, then that it is an IDE setting in VBx makes _somehow_ sense as only in the IDE there is a debugger.

So I like your attributes naming as this is very clear what's meant.

That said I don't know how complicated it get's once tB supports multi-project environments (workgroup) which each having a different error debugger setting. (?)

**Reply by bclothier** [reply]
I think the simplest approach is to treat the referenced project as if they are compiled since it's not typical to want to step through the referenced project's code. I think that if I want to step through both projects to debug the interpretability of the projects, I would need to do something explicitly to step through the referenced project. Within that project, the scope changes and we respect the current project's attributes.

**Reply by Kr00l** [reply]
I wouldn't agree to treat a referenced project as "runtime" behavior.

Typical use-case for a workgroup is to have a std-exe together with an ax-dll project so that it's running regfree in IDE.
When the std-exe calls the ax-dll you may want to debug the line failing in the ax-dll (break in class module option) and not on the calling side in the std-exe.

Another question:
Should `[DebuggerBehavior(MatchRuntimeBehavior)]` be the default for tB projects? If somebody complaints Wayne can argue it's not VB6 nor VBA   default. :)
